File: server/models/Lead.js
const mongoose = require('mongoose');

const statusHistorySchema = new mongoose.Schema({
    status: { type: String, required: true },
    changed_by: { type: String, default: 'system' },
    changed_by_name: { type: String, default: 'System' },
    catatan: { type: String, default: '' },
    changed_at: { type: Date, default: Date.now }
}, { _id: false });

const leadSchema = new mongoose.Schema({
    _id: { type: String }, // WA number atau UUID sebagai ID
    user_id: { type: String },
    nama_lengkap: { type: String, required: true },
    whatsapp_num: { type: String, required: true },
    domisili: { type: String, default: '' },
    yang_berangkat: { type: String, default: '' },
    paket_pilihan: { type: String, default: '' },
    kesiapan_paspor: { type: String, default: '' },
    fasilitas_utama: { type: String, default: '' },
    utm_source: { type: String, default: '' },
    utm_medium: { type: String, default: '' },
    utm_campaign: { type: String, default: '' },
    landing_page: { type: String, default: '' },
    form_source: { type: String, default: '' },
    status_followup: { type: String, default: 'New Data' },
    catatan: { type: String, default: '' },
    revenue: { type: Number, default: 0 },
    program_id: { type: String, default: '' },
    last_contact: { type: Date },
    rencana_umrah: { type: String, default: '' },
    assigned_to: { type: String, default: '' }, // username of assigned sales
    assigned_to_name: { type: String, default: '' },
    // Status change history
    status_history: { type: [statusHistorySchema], default: [] },
    is_deleted: { type: Boolean, default: false },
    deleted_at: { type: Date, default: null },
    is_restored: { type: Boolean, default: false },
}, {
    timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' }
});

module.exports = mongoose.model('Lead', leadSchema);

File: backend/internal/domain/lead.go
package domain

import (
	"context"
	"time"
)

type StatusHistory struct {
	Status        string    `json:"status"`
	ChangedBy     string    `json:"changed_by"`
	ChangedByName string    `json:"changed_by_name"`
	Catatan       string    `json:"catatan"`
	ChangedAt     time.Time `json:"changed_at"`
}

type Lead struct {
	ID               string                 `json:"id"`
	UserID           string                 `json:"user_id"`
	NamaLengkap      string                 `json:"nama_lengkap"`
	WhatsappNum      string                 `json:"whatsapp_num"`
	Domisili         string                 `json:"domisili"`
	YangBerangkat    string                 `json:"yang_berangkat"`
	JamaahUsiaDetail interface{}            `json:"jamaah_usia_detail"`
	PaketPilihan     string                 `json:"paket_pilihan"`
	KesiapanPaspor   string                 `json:"kesiapan_paspor"`
	PengalamanUmrah  string                 `json:"pengalaman_umrah"`
	FasilitasUtama   string                 `json:"fasilitas_utama"`
	UTMSource        string                 `json:"utm_source"`
	UTMMedium        string                 `json:"utm_medium"`
	UTMCampaign      string                 `json:"utm_campaign"`
	LandingPage      string                 `json:"landing_page"`
	FormSource       string                 `json:"form_source"`
	StatusFollowUp   string                 `json:"status_followup"`
	Catatan          string                 `json:"catatan"`
	Revenue          int64                  `json:"revenue"`
	ProgramID        string                 `json:"program_id"`
	LastContact      *time.Time             `json:"last_contact"`
	RencanaUmrah     string                 `json:"rencana_umrah"`
	AssignedTo       string                 `json:"assigned_to"`
	AssignedToName   string                 `json:"assigned_to_name"`
	StatusHistory    []StatusHistory        `json:"status_history"`
	IsDeleted        bool                   `json:"is_deleted"`
	DeletedAt        *time.Time             `json:"deleted_at"`
	Preferences      map[string]interface{} `json:"preferences"`
	CreatedAt        time.Time              `json:"created_at"`
	UpdatedAt        time.Time              `json:"updated_at"`
}

type LeadStats struct {
	TotalLeads     int                      `json:"totalLeads"`
	TodaysLeads    int                      `json:"todaysLeads"`
	TotalClosing   int                      `json:"totalClosing"`
	ConversionRate string                   `json:"conversionRate"`
	ChartData      []map[string]interface{} `json:"chartData"`
}

type LeadRepository interface {
	GetLeads(ctx context.Context, status, search, assignedTo string, limit, offset int) ([]Lead, int, error)
	GetLeadByID(ctx context.Context, id string) (*Lead, error)
	CreateLead(ctx context.Context, lead *Lead) error
	UpdateLead(ctx context.Context, lead *Lead) error
	DeleteLead(ctx context.Context, id string) error
	GetStats(ctx context.Context) (*LeadStats, error)
}

type LeadUsecase interface {
	FetchLeads(ctx context.Context, status, search, assignedTo string, limit, offset int) ([]Lead, int, error)
	GetLeadByID(ctx context.Context, id string) (*Lead, error)
	FetchLeadHistory(ctx context.Context, id string) ([]StatusHistory, error)
	AddLead(ctx context.Context, lead *Lead) error
	AddManualLead(ctx context.Context, lead *Lead, creatorUsername, creatorName string) error
	UpdateLead(ctx context.Context, id string, updates map[string]interface{}, updaterUsername, updaterName string) error
	DeleteLead(ctx context.Context, id string) error
	GetStatsOverview(ctx context.Context) (*LeadStats, error)
}
